analysis/05_hyperparameters.md

```diff
@@ -0,0 +1,187 @@
+anom = pa.setup(
+    data=df_idmc_model,
+    silent=True,
+    ignore_features=["id"],
+)
+df_results = pd.DataFrame()
+fractions = [0.005, 0.01, 0.02, 0.03, 0.04, 0.05]
+for fraction in fractions:
+    anom_model = pa.create_model(model=MODEL, fraction=fraction)
+    results = pa.assign_model(anom_model)
+    results["fraction"] = fraction
+    df_results = pd.concat([df_results, results], ignore_index=True)
+
+df_final = df_results.merge(
+    df_idmc,
+    on="id",
+    how="left",
+    copy=False,
+    suffixes=[None, "_todrop"],
+)
+df_final = df_final[
+    df_final.columns.drop(list(df_final.filter(regex="_todrop")))
+]
+```
+
+```python
+# Plot results for Myanmar
+chart1 = (
+    alt.Chart(df_final[df_final["iso3"] == "MMR"])
+    .mark_point(opacity=1)
+    .encode(
+        x="displacement_date",
+        y="fraction:O",
+        color=alt.Color(
+            "figure", scale=alt.Scale(type="log", scheme="orangeRed")
+        ),
+    )
+    .transform_filter((datum.Anomaly == 1))
+)
+
+chart2 = (
+    alt.Chart(df_final[df_final["iso3"] == "MMR"])
+    .mark_point(opacity=0.05)
+    .encode(
+        x="displacement_date",
+        y="fraction:O",
+    )
+    .transform_filter((datum.Anomaly == 0))
+)
+
+lc = chart2 + chart1
+
+lc.configure_axisX(labelAngle=90)
+```
+
+We can see that for Myanmar, we would need a fraction of at least
+3% to have detected the anomaly in April, and 2% for July.
+
+## K (number of nearest neighbours)
+
+```python
+# Run for different fractions
+anom = pa.setup(
+    data=df_idmc_model,
+    silent=True,
+    ignore_features=["id"],
+)
+df_results = pd.DataFrame()
+ks = [2, 3, 4, 5, 10, 20, 30]
+for k in ks:
+    anom_model = pa.create_model(model=MODEL, n_neighbors=k)
+    results = pa.assign_model(anom_model)
+    results["k"] = k
+    df_results = pd.concat([df_results, results], ignore_index=True)
+
+df_final = df_results.merge(
+    df_idmc,
+    on="id",
+    how="left",
+    copy=False,
+    suffixes=[None, "_todrop"],
+)
+df_final = df_final[
+    df_final.columns.drop(list(df_final.filter(regex="_todrop")))
+]
+```
+
+```python
+# Plot results for Myanmar
+chart1 = (
+    alt.Chart(df_final[df_final["iso3"] == "MMR"])
+    .mark_point(opacity=1)
+    .encode(
+        x="displacement_date",
+        y="k:O",
+        color=alt.Color(
+            "figure", scale=alt.Scale(type="log", scheme="orangeRed")
+        ),
+    )
+    .transform_filter((datum.Anomaly == 1))
+)
+
+chart2 = (
+    alt.Chart(df_final[df_final["iso3"] == "MMR"])
+    .mark_point(opacity=0.05)
+    .encode(
+        x="displacement_date",
+        y="k:O",
+    )
+    .transform_filter((datum.Anomaly == 0))
+)
+
+lc = chart2 + chart1
+
+lc.configure_axisX(labelAngle=90)
+```
+
+Hm so now the question, how do you tune hyperparameters
+for unsupervised anomaly detection? Briefly checking the
+literature I found a paper
+[here](https://github.com/albertcthomas/anomaly_tuning),
+to look into later. Will leave k at 5 for now.
```

Comment:
Hmm very good question to which I sadly don't have an answer either. The paper looks good very interesting. 
However, I would also wonder how much time you should spend tuning the hyperparameters..